pointcloud/pointcloud_file.go

```diff
@@ -502,3 +506,215 @@ func readSliceToPoint(slice []float64, header pcdHeader) (r3.Vector, Data, error
 	}
 }
+
+// reorganizeToStructureOfArrays converts point cloud data from array-of-structures
+// to structure-of-arrays format for better compression.
+func reorganizeToStructureOfArrays(cloud PointCloud) ([]byte, error) {
+	size := cloud.Size()
+	if size == 0 {
+		return nil, errors.New("empty point cloud")
+	}
+
+	hasColor := cloud.MetaData().HasColor
+	var data []byte
+
+	if hasColor {
+		// Reserve space for x, y, z, rgb arrays
+		data = make([]byte, 0, size*16) // 4 float32s per point
+
+		// Separate arrays for x, y, z, rgb
+		xData := make([]byte, 0, size*4)
+		yData := make([]byte, 0, size*4)
+		zData := make([]byte, 0, size*4)
+		rgbData := make([]byte, 0, size*4)
+
+		cloud.Iterate(0, 0, func(pos r3.Vector, d Data) bool {
+			// Convert RDK units (millimeters) to meters for PCD
+			x := pos.X / 1000.
+			y := pos.Y / 1000.
+			z := pos.Z / 1000.
+
+			buf := make([]byte, 4)
+			binary.LittleEndian.PutUint32(buf, math.Float32bits(float32(x)))
+			xData = append(xData, buf...)
+
+			binary.LittleEndian.PutUint32(buf, math.Float32bits(float32(y)))
+			yData = append(yData, buf...)
+
+			binary.LittleEndian.PutUint32(buf, math.Float32bits(float32(z)))
+			zData = append(zData, buf...)
+
+			c := _colorToPCDInt(d)
+			binary.LittleEndian.PutUint32(buf, uint32(c))
+			rgbData = append(rgbData, buf...)
+
+			return true
+		})
+
+		// Combine all arrays
+		data = append(data, xData...)
+		data = append(data, yData...)
+		data = append(data, zData...)
+		data = append(data, rgbData...)
+	} else {
+		// Reserve space for x, y, z arrays
+		data = make([]byte, 0, size*12) // 3 float32s per point
+
+		// Separate arrays for x, y, z
+		xData := make([]byte, 0, size*4)
+		yData := make([]byte, 0, size*4)
+		zData := make([]byte, 0, size*4)
+
+		cloud.Iterate(0, 0, func(pos r3.Vector, d Data) bool {
+			// Convert RDK units (millimeters) to meters for PCD
+			x := pos.X / 1000.
+			y := pos.Y / 1000.
+			z := pos.Z / 1000.
+
+			buf := make([]byte, 4)
+			binary.LittleEndian.PutUint32(buf, math.Float32bits(float32(x)))
+			xData = append(xData, buf...)
+
+			binary.LittleEndian.PutUint32(buf, math.Float32bits(float32(y)))
+			yData = append(yData, buf...)
+
+			binary.LittleEndian.PutUint32(buf, math.Float32bits(float32(z)))
+			zData = append(zData, buf...)
+
+			return true
+		})
+
+		// Combine all arrays
+		data = append(data, xData...)
+		data = append(data, yData...)
+		data = append(data, zData...)
+	}
+
+	return data, nil
+}
+
+// writePCDCompressed writes compressed point cloud data using LZF compression.
+func writePCDCompressed(cloud PointCloud, out io.Writer) error {
+	// Reorganize data to structure-of-arrays format
+	uncompressedData, err := reorganizeToStructureOfArrays(cloud)
+	if err != nil {
+		return err
+	}
+
+	// Compress the data using LZF
+	// Allocate output buffer with maximum possible size
+	compressedData := make([]byte, len(uncompressedData)+len(uncompressedData)/64+16+3)
```

Comment:
I don't understand this line and what these magic numbers are doing.  It seems that the maximum size of the compressed data would be the size of the uncompressed data, no?